core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -127,20 +140,137 @@ public Response deleteBucket(@PathParam("bucket") final String bucket) {
    */
+  @PUT
+  @Path(OBJECT_PARAM)
+  @ReturnType("java.lang.Void")
+  @Consumes(MediaType.APPLICATION_OCTET_STREAM)
+  public Response createObject(@HeaderParam("Content-MD5") final String contentMD5,
+                               @PathParam("bucket") final String bucket,
+                               @PathParam("object") final String object,
+                               final InputStream is) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response>() {
+      @Override
+      public Response call() throws S3Exception {
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+        checkBucketIsAlluxioDirectory(bucketPath);
+        AlluxioURI objectURI = new AlluxioURI(bucketPath + AlluxioURI.SEPARATOR + object);
+
+        try {
+          CreateFileOptions options = CreateFileOptions.defaults().setRecursive(true)
+              .setWriteType(getS3WriteType());
+          FileOutStream os = mFileSystem.createFile(objectURI, options);
+          MessageDigest md5 = MessageDigest.getInstance("MD5");
+
+          byte[] buf = new byte[BUFFER_SIZE];
+          while (true) {
+            int len = is.read(buf);
+            if (len == -1) {
+              break;
+            }
+            md5.update(buf, 0, len);
+            os.write(buf, 0, len);
+          }
+          os.close();
+
+          byte[] digest = md5.digest();
+          String base64Digest = BaseEncoding.base64().encode(digest);
+          if (!contentMD5.equals(base64Digest)) {
+            // The object may be corrupted, delete the written object and return an error.
+            mFileSystem.delete(objectURI);
+            throw new S3Exception(objectURI.getPath(), S3ErrorCode.BAD_DIGEST);
+          }
+
+          String entityTag = Hex.encodeHexString(digest);
+          return Response.ok().tag(entityTag).build();
+        } catch (Exception e) {
+          throw toObjectS3Exception(e, bucketPath);
+        }
+      }
+    });
+  }
+
+  /**
+   * @summary retrieves an object's metadata
+   * @param bucket the bucket name
+   * @param object the object name
+   * @return the response object
+   */
+  @HEAD
+  @Path(OBJECT_PARAM)
+  @ReturnType("java.lang.Void")
+  public Response getObjectMetadata(@PathParam("bucket") final String bucket,
+                            @PathParam("object") final String object) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response>() {
+      @Override
+      public Response call() throws S3Exception {
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+        checkBucketIsAlluxioDirectory(bucketPath);
+        AlluxioURI objectURI = new AlluxioURI(bucketPath + AlluxioURI.SEPARATOR + object);
+
+        try {
+          URIStatus status = mFileSystem.getStatus(objectURI);
+          // TODO(cc): Consider how to respond with the object's ETag.
+          return Response.ok().lastModified(new Date(status.getLastModificationTimeMs())).build();
```

Comment:
create a utility function for uri status to response? (in case we decide to add more information)